File: challenges/related_questions.py
class Tree(object):
	"""docstring for Tree."""
	def __init__(self,n,nodes):
		self.no_of_nodes = n
		self.root = None
		self.nodes = nodes
		self.adjacency_list = {}
		self.optimum_root = None
		self.visit = []

	def set_root(self):
	 	branches = [[x[0],len(x[1])] for x in self.adjacency_list.items()]
	#	print "branches ", branches
		max_root = max(branches, key = lambda x : x[1])
	#	print max_root
		self.root = max_root[0]


	def add_edge(self,x,y):
		if x in self.adjacency_list:
			self.adjacency_list[x].append(y)
		else:
			self.adjacency_list[x] = [y]

		if y in self.adjacency_list:
			self.adjacency_list[y].append(x)
		else:
			self.adjacency_list[y] = [x]

	def dfs(self,node_p,root_branch_no,sum_parent):
		adj_list = self.adjacency_list[node_p]
		count = 0
		print "node ", node_p, " childs ", adj_list
		self.nodes[node_p].parent_path_sum = sum_parent
		self.nodes[node_p].parent_branch_index = root_branch_no
		#print "node ", node_p, " childs ", adj_list
		for ch in adj_list:
			if self.visit[ch] == 0:
				count = count + 1
				self.visit[ch] = 1
				sum_branch = self.dfs(ch,root_branch_no,sum_parent+self.nodes[node_p].cost)
				self.nodes[node_p].children_path_sum.append(sum_branch)
		count = max(1,count)
		return count*self.nodes[node_p].cost + sum(self.nodes[node_p].children_path_sum)

	def traversal(self):
		self.visit = [0 for i in range(self.no_of_nodes)]
		adj_list = self.adjacency_list[self.root]
		#print adj_list
		branch_count = 0
		self.visit[self.root] = 1
		for ch in adj_list:
			if self.visit[ch] == 0:
				branch_count += 1
				self.visit[ch] = 1
				sum_branch = self.dfs(ch,branch_count,self.nodes[self.root].cost)
				self.nodes[self.root].children_path_sum.append(sum_branch)
		return

	def print_graph(self):
		print "root index ", self.root
		print "root branch sums ", self.nodes[self.root].children_path_sum
		print "Node info"
		for i in range(1,self.no_of_nodes):
			print "node no ", self.nodes[i].ind," parent_br ", self.nodes[i].parent_branch_index, " path_sum ", self.nodes[i].parent_path_sum

	 	return

class Node(object):
	"""docstring for Node."""
	def __init__(self,q_id,weight):
		self.ind = q_id
		self.cost = weight
		self.children_path_sum = []
		self.parent_path_sum = 0
		self.parent_branch_index = -1

def main():
	n = int(raw_input())
	cost = map(int,raw_input().split())
	questions = []
	questions.append(Node(0,0))
	for i in range(1,n+1):
		node_m = Node(i,cost[i-1])
		questions.append(node_m)

	#for i in range(len(questions)):
	#	print questions[i].ind

	#Construct graph
	graph_of_question = Tree(n+1,questions)

	# add edges
	for i in range(n-1):
		edge = map(int,raw_input().split())
		graph_of_question.add_edge(edge[0],edge[1])

	# set default root
	graph_of_question.set_root()
	#print graph_of_question.root
	graph_of_question.traversal()
	graph_of_question.print_graph()
	#build cost matrix
	#find optimum root
	#root = graph_of_question.find_optimum_root()

	#print root+1

if __name__ == "__main__":
	main()

File: challenges/related_questions.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <vector>

using namespace std;

bool visit[100001];

struct node {
    int id;
    double weight;
    vector<node*> v;
    long double total_weight;
    long int child_count;
    bool is_child;

    node(int i, int w){
        id = i;
        weight = w;
        is_child = 0;
        child_count = 0;
        total_weight = 0;
    }
};

long double dfs(node* root){
    long double sum;
    int count = 0;
    for(int i = 0; i < root->v.size();i++){
        if (visit[root->v[i]->id] == 0){
            visit[root->v[i]->id] = 1;
            count++;
            sum = dfs(root->v[i]);
            root->total_weight +=sum;
        }
    }
    //cout << root->id << " " << root->total_weight << endl;
    if (count == 0) {
        root->total_weight += root->weight;
    }
    else{
        root->total_weight = (root->total_weight/count) + root->weight;
    }
    sum = root->total_weight;
    return sum;
}
void traverse_util(vector<node*> graph, int id, int size) {
    int n = size;
    for(int i = 1; i <= n; i++){
        visit[i] = 0;
    }
    visit[id] = 1;
    long double sum = dfs(graph[id]);
//    for (int i = 1; i <= n; i++) {
//        cout << graph[i]->id << " " << graph[i]->total_weight << endl;
//    }
    for (int i = 1; i <= n; i++){
        visit[i] = 0;
    }
    queue<node*> q;
    visit[id] = 1;
    q.push(graph[id]);
    node* root = NULL;
    node* temp = NULL;
    long double current_weight = 0;
    long double current_count = 0;
    while(q.size() != 0) {
        root = q.front();
        q.pop();
        for(int i = 0; i < root->v.size(); i++){
            if (visit[root->v[i]->id] == 0){
                temp = root->v[i];
                visit[temp->id] = 1;
                current_weight = root->total_weight - root->weight;
                current_count = (double)root->v.size();
                if (current_count > 1) {
                    current_weight = current_weight - (temp->total_weight/current_count);
                    current_weight = current_weight*(current_count/(current_count-1)) + root->weight;
                    temp->total_weight = ((temp->total_weight - temp->weight)*(temp->v.size()-1) + current_weight)/(double)temp->v.size() + temp->weight;
                }
                q.push(temp);
            }
        }
    }
//    for (int i = 1; i <= n; i++) {
//        cout << graph[i]->id << " " << graph[i]->total_weight << endl;
//    }

    double estimate = 0.0;
    double min = graph[1]->total_weight;
    int index = 1;
    for (int i = 2; i <= n; i++) {
        estimate = graph[i]->total_weight;
        if (estimate < min ){
            min = estimate;
            index = i;
        }
    }
    cout << index << endl;
    return;
}
int main(){
    int n;
    int weight;
    cin >> n;
    node *question;
    vector<node*> questions(n+1);
    for (int i = 1; i <= n; i++){
        cin >> weight;
        questions[i] = new node(i,weight);
    }
    int q1,q2;
    for(int i = 1; i < n; i++){
        cin >> q1 >> q2;
        questions[q1]->v.push_back(questions[q2]);
        questions[q2]->v.push_back(questions[q1]);
    }
    int temp_count = 0;
    int temp_id = 0;
    for(int i = 1; i <= n; i++) {
        if (questions[i]->v.size() == 1){
            questions[i]->child_count = 1;
            questions[i]->is_child = 1;
        }
        if (questions[i]->v.size() > temp_count){
            temp_count = questions[i]->v.size();
            temp_id = i;
        }
    }
//    cout << "root " << temp_id << " " << temp_count << endl;
    traverse_util(questions,temp_id,n);
    return 0;
}
